# Project: Data Analysis with Pandas

In this project, you will work with a dataset containing information about movies and shows. You will use the pandas library to read, clean, and analyze the data. This project will help you practice working with DataFrames, indexing, and data manipulation using pandas.


## Dataset Description

The dataset `movies_and_shows.csv` contains information about various movies and shows, including:

- **name**: The name of the actor or actress.
- **Character**: The character they played.
- **r0le**: The role type (e.g., ACTOR).
- **TITLE**: The title of the movie or show.
- **Type**: Whether it's a MOVIE or SHOW.
- **release Year**: The year it was released.
- **genres**: A list of genres the movie or show belongs to.
- **imdb sc0re**: The IMDb score of the movie or show.
- **imdb v0tes**: The number of votes on IMDb.

Here's a small sample of the dataset:

| name              | Character                         | r0le   | TITLE        |  Type | release Year | genres             | imdb sc0re | imdb v0tes |
|-------------------|-----------------------------------|--------|--------------|-------|--------------|--------------------|------------|------------|
| Robert De Niro    | Travis Bickle                     | ACTOR  | Taxi Driver  | MOVIE | 1976         | ['drama', 'crime'] | 8.2        | 808582     |
| Jodie Foster      | Iris Steensma                     | ACTOR  | Taxi Driver  | MOVIE | 1976         | ['drama', 'crime'] | 8.2        | 808582     |
| Albert Brooks     | Tom                               | ACTOR  | Taxi Driver  | MOVIE | 1976         | ['drama', 'crime'] | 8.2        | 808582     |
| Harvey Keitel     | Matthew 'Sport' Higgins           | ACTOR  | Taxi Driver  | MOVIE | 1976         | ['drama', 'crime'] | 8.2        | 808582     |
| Cybill Shepherd   | Betsy                             | ACTOR  | Taxi Driver  | MOVIE | 1976         | ['drama', 'crime'] | 8.2        | 808582     |



## Getting Started

Let's begin by setting up our environment.

### Importing Libraries

First, we need to import the necessary libraries.

**Instructions:**

- Import the `pandas` library, which we will use for data manipulation.

In [4]:
# Import the pandas library as 'pd'
import pandas as pd

We will load the dataset into a pandas DataFrame to begin our analysis.

**Instructions:**

- Read the `movies_and_shows.csv` file into a DataFrame. It is in the /datasets directory so the full path to include in the read_csv method will be "/datasets/movies_and_shows.csv"
- Display the first few rows of the DataFrame to get an initial look at the data.

**Hints:**

- Use `pd.read_csv()` to read the CSV file and store it in a variable called "df".
- Use the `.head()` method to display the first few rows.

In [5]:
# Read the dataset into a DataFrame
df = pd.read_csv('/datasets/movies_and_shows.csv')

In [6]:
# Display the first few rows of the DataFrame
df.head()

name                Character   r0le        TITLE   Type  \
0   Robert De Niro            Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster            Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                      Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                    Betsy  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0

Understanding the structure and content of your data is crucial before performing any analysis.

**Instructions:**

- Use the `.info()` method to get information about the DataFrame.
- Identify any issues with the Column names.

In [7]:
# Get information about the DataFrame and print the column names
df.info

<bound method DataFrame.info of                      name                Character      r0le        TITLE  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
85574    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
85575  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
85576        In??s Prieto                    Fanny     ACTOR      Lokillo   
85577        Isabel Gaona                   Cacica     ACTOR      Lokillo   
85578      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            Type  release Year             

## Task 1: Data Cleaning

Let's clean the data to fix issues with the column names

**Instructions:**
- Rename the columns to correct any errors and make them consistent.

**Hints:**

- Use the `.rename()` method to rename columns.
- Pass a dictionary to the `columns` parameter of `.rename()`, where the keys are the old column names and the values are the new names.
- Remove unnecessary whitespace from the column names.

In [8]:
# Rename columns to make them consistent and correct errors
df = df.rename(columns= {'   name':'name', 'Character':'character', 'r0le':'role', 'TITLE':'title', '  Type':'type', 'release Year':'release_year', 'imdb sc0re':'imdb_score', 'imdb v0tes':'imdb_votes'})

In [9]:
# Print the updated column names to confirm changes
print(df.columns)

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')


## Task 2: Correcting a Misspelled Name in the Data

While analyzing the dataset, you notice that some names are misspelled or contain special characters due to encoding issues. Accurate data is essential for reporting and recommendations, so let’s correct one of these entries.

### Instructions

1. **Locate the Row with the Incorrect Name**:
   - Use `.loc[]` to retrieve the row where `name` is `"In??s Prieto"`.
   - You can locate a row based on the index (85576) and column name called "name".
   - Print the row to verify that you have the correct one.


2. **Correct the Name**:
   - Using `.loc[]`, update the `name` column for this row to "Ines Prieto."
   
3. **Verify the Correction**:
   - Print the row again to ensure that the name has been corrected.

In [10]:
# Locate the row with the incorrect name
row = df.loc[df['name']=="In??s Prieto"]
print(row)

               name character   role    title       type  release_year  \
77798  In??s Prieto     Fanny  ACTOR  Lokillo      MOVIE          2021   
85576  In??s Prieto     Fanny  ACTOR  Lokillo  the movie          2021   

           genres  imdb_score  imdb_votes  
77798  ['comedy']         3.8        68.0  
85576  ['comedy']         3.8        68.0  


In [11]:
# Correct the name
df.loc[df['name']=="In??s Prieto", 'name'] = 'Ines Prieto'

In [12]:
# Verify the correction
print(df.loc[df['name']=="In??s Prieto"])
print(df.loc[df['name']=="Ines Prieto"])

Empty DataFrame
Columns: [name, character, role, title, type, release_year, genres, imdb_score, imdb_votes]
Index: []
              name character   role    title       type  release_year  \
77798  Ines Prieto     Fanny  ACTOR  Lokillo      MOVIE          2021   
85576  Ines Prieto     Fanny  ACTOR  Lokillo  the movie          2021   

           genres  imdb_score  imdb_votes  
77798  ['comedy']         3.8        68.0  
85576  ['comedy']         3.8        68.0  


## Task 3: Finding All Movies and Shows Featuring Ines Prieto

Now that we've corrected the spelling of "Ines Prieto" in the dataset, let's find all the TV shows and movies she has acted in. This type of filtering is helpful for generating actor-specific profiles or building a list of their works.

### Instructions

1. **Filter by Actor’s Name**:
   - Use a filtering condition to select rows where the `name` column is equal to `"Ines Prieto"`.
   
2. **Display Relevant Columns**:
   - From each matching row, retrieve only the `title`, `release_year`, `imdb_score`, and `genres` columns for a clear, concise output.

**Hint:**

To filter rows based on a specific value in a column, use a condition inside df[ ... ]. In this case, check if the name column equals "Ines Prieto". Then, select only the columns you need (like title, release_year, imdb_score, and genres) by specifying them in double brackets [ [ ... ] ].


In [13]:
# Filter rows where the actor's name is "Ines Prieto" and display the title, release_year, genres, and imdb_score
filtered_rows = df.loc[df['name'] == 'Ines Prieto', ['title', 'release_year', 'genres', 'imdb_score']]

In [14]:
# Display the result
print(filtered_rows)

         title  release_year      genres  imdb_score
77798  Lokillo          2021  ['comedy']         3.8
85576  Lokillo          2021  ['comedy']         3.8


## Task 4: Finding Highly Rated Movies

We want to identify movies with an IMDb rating of at least **9.0**. This list could be helpful for curating a "Top Movies" section based on high ratings.

### Instructions

1. **Filter for High IMDb Scores**:
   - First, filter the DataFrame to include only rows where the `imdb_score` is greater than 9.0.

2. **Extract the Titles**:
   - From this filtered DataFrame, select only the `title` column, which contains the names of the movies.

3. **Get Unique Titles**:
   - Convert the resulting list of titles to a set to remove any duplicate titles. Using `set()` will keep only unique movie names.

   - *Example*: `unique_titles = set(high_score_titles)`

4. **Print the Unique Titles**:
   - Display the final set of unique movie titles to see the list of top-rated movies.


In [15]:
# Filter for movies with an IMDb score above 9.0
movies_above_9 = df.loc[df['imdb_score'] > 9.0]

# Extract the 'title' column from the filtered DataFrame
titles_of_movies_above_9 = movies_above_9.loc[:,'title']

# Get a unique set of titles
unique_titles_of_movies_above_9 = set(titles_of_movies_above_9)

# Print the unique titles
print(unique_titles_of_movies_above_9)

{'Our Planet', 'Breaking Bad', 'The Last Dance', 'Major', 'Reply 1988', 'Avatar: The Last Airbender', 'My Mister', 'Kota Factory'}


## Task 5: Creating a Function to Find Unique Top-Rated Movies

In this task, we’ll create a function to find unique movies with an IMDb score above a certain threshold that the user provides.

### Instructions

1. **Define the Function**:
   - Start by creating a function called `get_unique_top_movies` that takes one parameter, `min_score`.

2. **Filter the Data**:
   - Inside the function, create a new variable (e.g., `high_score_df`) that stores rows where the `imdb_score` is greater than or equal to `min_score`.

3. **Extract Movie Titles**:
   - In a new variable (e.g., `high_score_titles`), select the `title` column from the filtered DataFrame.

4. **Remove Duplicate Titles**:
   - Convert `high_score_titles` to a set using `set(high_score_titles)` to automatically remove duplicates, ensuring each title appears only once.

5. **Return the Unique Titles**:
   - Make sure the function returns the `unique_titles` set.


In [16]:
# Define the function
def get_unique_top_movies(min_score):
    # Filter for movies with IMDb score above min_score
    filtered_movies = df.loc[df['imdb_score'] > min_score]

    # Extract the 'title' column
    titles_filtered = filtered_movies.loc[:,'title']

    # Remove duplicate titles
    unique_titles_filtered = set(titles_filtered)

    # Return unique titles
    return unique_titles_filtered

In [17]:
# Test the function
print(get_unique_top_movies(9.0))

{'Our Planet', 'Breaking Bad', 'The Last Dance', 'Major', 'Reply 1988', 'Avatar: The Last Airbender', 'My Mister', 'Kota Factory'}


## Task 6: Creating a Function to Find Top Movies from a Specific Decade

Let’s create a function to retrieve movies from a particular decade with high IMDb ratings. This function can be useful for generating lists of top-rated movies from different time periods.

### Instructions

1. **Define the Function**:
   - Create a function called `get_top_movies_from_decade` that accepts `decade_start` and `min_score` as parameters.

2. **Filter by Decade**:
   - Inside the function, filter the DataFrame to include only movies where `release_year` is within the specified decade.
   - *Hint*: You can achieve this by checking that `release_year` is between `decade_start` and `decade_start + 9`

3. **Filter by IMDb Score**:
   - Further filter this subset to include only movies where `imdb_score` is greater than or equal to `min_score`.

4. **Extract Movie Titles**:
   - From the resulting DataFrame, select the `title` column and remove duplicates using set().

5. **Return the List of Titles**:
   - Return the set of the best movies from that decade.

By completing this task, you’ll have a function that returns unique, top-rated movies from a specified decade. This is useful for highlighting standout movies from different eras, perfect for “Best of the Decade” features!

In [18]:
# Define the function
def get_top_movies_from_decade(decade_start, min_score):
    # Filter for movies released within the decade
    first_filter = df[(df['release_year'] >= decade_start) 
        & (df['release_year'] <= decade_start + 9)] # Included 1999

    # Further filter by IMDb score
    second_filter = first_filter.loc[first_filter['imdb_score'] >= min_score]
    titles_second_filter = second_filter.loc[:, 'title']

    # Extract and remove duplicate titles
    unique_filtered_movies = set(titles_second_filter)

    # Return unique titles
    return unique_filtered_movies

In [19]:
# Test the function
print(get_top_movies_from_decade(1990, 8.5))

{'Se7en', 'GoodFellas', 'Bill Hicks: Revelations', 'Cowboy Bebop', 'Neon Genesis Evangelion', 'Forrest Gump', 'One Piece', 'L??on: The Professional'}



## Task 7: Creating a Function to List All Actors in a Given Title

Imagine you want to list all the actors in a specific movie or show. Let’s create a function that takes a `title` as input and returns the names of all actors in that title, combined into a single string.


### Instructions

1. **Define the Function**:
   - Create a function called `get_actors_for_title` that accepts one parameter, `title`.
   
2. **Filter by Title and Role**:
   - Inside the function, filter the DataFrame to select rows where the `title` column matches the provided `title` parameter and the `role` column is `"ACTOR"` (to ensure you only retrieve actors).

3. **Extract Actor Names**:
   - From this filtered DataFrame, select only the `name` column to get the actor names.

4. **Combine Names into a Single String**:
   - Use `', '.join()` to combine the list of actor names into a single string, with each name separated by a comma.

5. **Return the Result**:
   - Return the resulting string of actor names.


This function will allow you to retrieve and format a list of actors for any movie or show title, which is useful for creating cast lists or displaying actors for specific titles.

In [20]:
# Define the function
def get_actors_for_title(title):
    
    # Filter for rows with the specified title and role as 'ACTOR'
    filtered_rows = df.loc[df['title'] == title]
    filtered_rows_two = filtered_rows.loc[filtered_rows['role'] == 'ACTOR']

    # Extract the 'name' column for actor names
    names_column = filtered_rows_two.loc[:, 'name']


    # Combine names into a single string
    names_column = ', '.join(names_column)

    # Return the result
    return names_column

In [21]:
# Test the function
print(get_actors_for_title("Taxi Driver"))

Robert De Niro, Jodie Foster, Albert Brooks, Harvey Keitel, Cybill Shepherd, Peter Boyle, Leonard Harris, Diahnne Abbott, Gino Ardito, Martin Scorsese, Murray Moston, Richard Higgs, Bill Minkin, Bob Maroff, Victor Argo, Joe Spinell, Robinson Frank Adu, Brenda Dickson, Norman Matlock, Harry Northup, Harlan Cary Poe, Steven Prince, Peter Savage, Nicholas Shields, Ralph S. Singleton, Annie Gagen, Carson Grant, Mary-Pat Green, Debbi Morgan, Don Stroud, Copper Cunningham, Garth Avery, Nat Grant, Billie Perkins, Catherine Scorsese, Charles Scorsese, Odunlade Adekola, Ijeoma Grace Agu


## Task 8: Creating a Function to Categorize Movies by IMDb Score

Let’s categorize movies and shows based on their IMDb scores to provide a quick evaluation of their popularity or quality. We’ll create a function that takes in a `title` and returns a rating category based on the movie or show’s IMDb score.

### Rating Categories

- **Excellent**: IMDb score of 9.0 or higher
- **Good**: IMDb score between 7.0 and 8.9
- **Average**: IMDb score between 5.0 and 6.9
- **Low**: IMDb score below 5.0

### Instructions

1. **Define the Function**:
   - Create a function called `categorize_imdb_score` that accepts one parameter, `title`.

2. **Filter by Title**:
   - Inside the function, filter the DataFrame to find the row where `title` matches the given title.

3. **Retrieve IMDb Score**:
   - If the title exists, retrieve the `imdb_score` for that movie or show.
   - If the title is not found, return `"Title not found"`.

4. **Categorize Score with `if-else`**:
   - Use `if-elif-else` statements to evaluate the `imdb_score` and return one of the rating categories based on the ranges provided.

5. **Return the Category**:
   - Return the appropriate category based on the score.

This function allows you to quickly categorize movies or shows by their IMDb score, which could be helpful for creating rating tags or recommending top-rated content.

In [24]:
# Define the function
def categorize_imdb_score(title):
    # Filter for the row with the specified title
    filtered_row = df.loc[df['title'] == title]
    imdb_score = 0
    
    # Check if title exists
    if title in df.values:
        # Retrieve the IMDb score for the movie
        imdb_score = filtered_row.loc[0, 'imdb_score'] # There's two different 'Taxi Driver' movies
    else:
        return 'Title not found'
    
    # Categorize score using if-else and return the ranking accordingly
    if imdb_score >= 9.0:
        return 'Excellent'
    elif imdb_score <= 8.9 and imdb_score >= 7.0:
        return 'Good'
    elif imdb_score <= 6.9 and imdb_score >= 5.0:
        return 'Average'
    else:
        return 'Low'


In [25]:
# Test the function
print(categorize_imdb_score("Taxi Driver")) # Should be 'Good'

Good


In [ ]:
# I don't see the comments for feedback?